discovery/kubernetes/client_metrics.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/metrics"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	cacheMetricsNamespace     = metricsNamespace + "_cache"
+	workqueueMetricsNamespace = metricsNamespace + "_workqueue"
+)
+
+var (
+	// Metrics for client-go's HTTP requests.
+	clientGoRequestResultMetricVec = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_total",
+			Help:      "Total number of HTTP requests to the Kubernetes API by status code.",
+		},
+		[]string{"status_code"},
+	)
+	clientGoRequestLatencyMetricVec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_duration_seconds",
+			Help:      "Histogram of latencies for HTTP requests to the Kubernetes API by endpoint.",
+		},
+		[]string{"endpoint"},
+	)
+
+	// Definition of metrics for client-go cache metrics provider.
+	clientGoCacheListTotalMetricVec = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_total",
+			Help:      "Total number of list operations.",
+		},
+		[]string{"cache_name"},
+	)
+	clientGoCacheListDurationMetricVec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_duration_seconds",
+			Help:      "Duration of a Kubernetes API call in seconds.",
+		},
+		[]string{"cache_name"},
+	)
+	clientGoCacheItemsInListCountMetricVec = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_items_count",
```

Comment:
Summary metrics should not end in _count, that's for the time series inside of them.